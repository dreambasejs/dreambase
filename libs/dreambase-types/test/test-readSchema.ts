import {
  Indexed,
  PrimaryKey,
  Type,
  parseType,
  readSchema,
  ArrayOf,
} from "../src/index";

describe("test-readSchema", () => {
  class Friend {
    id = PrimaryKey(Number, { autoIncrement: true });
    name = Type(String);
    age = Indexed(Number);
    tags = ArrayOf(Indexed(String));
    strArrWithDefault = ArrayOf(String, { default: ["foo", "bar", "3"] });
    cars = ArrayOf(Car);
    school = {
      name: Type(String),
      address: {
        city: Indexed(String, { compoundWith: () => this.school.name }),
      },
    };
  }

  class Car {
    name = Type(String);
    date = Type(Date);
  }

  it("Friend should match snapshot", () => {
    const friendType = readSchema(Friend);
    expect(friendType).toMatchSnapshot();
  });

  it("Car should match snapshot", () => {
    const friendType = readSchema(Car);
    expect(friendType).toMatchSnapshot();
  });
});
