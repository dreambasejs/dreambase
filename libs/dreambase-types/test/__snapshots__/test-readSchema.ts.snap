// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test-readSchema Car should match snapshot 1`] = `
Object {
  "ctor": [Function],
  "default": Car {
    "date": 1970-01-01T00:00:00.000Z,
    "name": "",
  },
  "properties": Object {
    "date": Object {
      "ctor": [Function],
      "default": 1970-01-01T00:00:00.000Z,
      "keyPath": "date",
      "parent": [Circular],
      "type": "Date",
      "typeSpecificOptions": Object {},
    },
    "name": Object {
      "ctor": [Function],
      "default": "",
      "keyPath": "name",
      "parent": [Circular],
      "type": "string",
      "typeSpecificOptions": Object {},
    },
  },
  "type": "Car",
  "typeSpecificOptions": Object {},
}
`;

exports[`test-readSchema Friend should match snapshot 1`] = `
Object {
  "ctor": [Function],
  "default": Friend {
    "age": 0,
    "id": 0,
    "name": "",
    "school": Object {
      "address": Object {
        "city": "",
      },
      "name": "",
    },
    "tags": Array [],
  },
  "properties": Object {
    "age": Object {
      "ctor": [Function],
      "default": 0,
      "indexes": Array [
        Object {
          "unique": false,
        },
      ],
      "keyPath": "age",
      "parent": [Circular],
      "type": "number",
      "typeSpecificOptions": Object {},
    },
    "id": Object {
      "ctor": [Function],
      "default": 0,
      "indexes": Array [
        Object {
          "autoIncrement": true,
          "isPrimaryKey": true,
          "unique": true,
        },
      ],
      "keyPath": "id",
      "parent": [Circular],
      "type": "number",
      "typeSpecificOptions": Object {},
    },
    "name": Object {
      "ctor": [Function],
      "default": "",
      "keyPath": "name",
      "parent": [Circular],
      "type": "string",
      "typeSpecificOptions": Object {},
    },
    "school": Object {
      "ctor": [Function],
      "default": Object {},
      "keyPath": "school",
      "parent": [Circular],
      "properties": Object {
        "address": Object {
          "ctor": [Function],
          "default": Object {},
          "keyPath": "school.address",
          "parent": [Circular],
          "properties": Object {
            "city": Object {
              "ctor": [Function],
              "default": "",
              "indexes": Array [
                Object {
                  "compoundKeys": Array [
                    Object {
                      "ctor": [Function],
                      "default": "",
                      "keyPath": "school.name",
                      "parent": [Circular],
                      "type": "string",
                      "typeSpecificOptions": Object {},
                    },
                  ],
                  "unique": false,
                },
              ],
              "keyPath": "school.address.city",
              "parent": [Circular],
              "type": "string",
              "typeSpecificOptions": Object {
                "compoundWith": [Function],
              },
            },
          },
          "type": "structure",
        },
        "name": Object {
          "ctor": [Function],
          "default": "",
          "keyPath": "school.name",
          "parent": [Circular],
          "type": "string",
          "typeSpecificOptions": Object {},
        },
      },
      "type": "structure",
    },
    "tags": Object {
      "ctor": [Function],
      "default": Array [],
      "item": Object {
        "ctor": [Function],
        "default": "",
        "indexes": Array [
          Object {
            "unique": false,
          },
        ],
        "keyPath": "tags",
        "parent": [Circular],
        "type": "string",
        "typeSpecificOptions": Object {},
      },
      "keyPath": "tags",
      "parent": [Circular],
      "type": "array",
    },
  },
  "type": "Friend",
  "typeSpecificOptions": Object {},
}
`;
